src/coreclr/jit/importer.cpp

```diff
@@ -2958,78 +2926,79 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken,
 
-                                    TypeCompareState castResult =
-                                        info.compCompHnd->compareTypesForCast(pResolvedToken->hClass,
-                                                                              isInstResolvedToken.hClass);
-                                    if (castResult != TypeCompareState::May)
-                                    {
-                                        JITDUMP("\n Importing BOX; ISINST; BR_TRUE/FALSE as constant\n");
-                                        impPopStack();
+                                TypeCompareState castResult =
+                                    info.compCompHnd->compareTypesForCast(pResolvedToken->hClass,
+                                                                          isInstResolvedToken.hClass);
 
-                                        impPushOnStack(gtNewIconNode((castResult == TypeCompareState::Must) ? 1 : 0),
-                                                       typeInfo(TYP_INT));
+                                if (castResult != TypeCompareState::May)
+                                {
+                                    JITDUMP("\n Importing BOX; ISINST; BR_TRUE/FALSE as constant\n");
 
-                                        // Skip the next isinst instruction
-                                        return 1 + sizeof(mdToken);
+                                    GenTree* op = impPopStack().val;
+                                    if ((op->gtFlags & GTF_SIDE_EFFECT) != 0)
+                                    {
+                                        impStoreTemp(lvaGrabTemp(true DEBUGARG("spill side effects")), op,
+                                                     CHECK_SPILL_ALL);
                                     }
+                                    impPushOnStack(gtNewIconNode((castResult == TypeCompareState::Must) ? 1 : 0),
+                                                   typeInfo(TYP_INT));
+                                    return 1 + sizeof(mdToken);
                                 }
-                                else if (foldAsHelper == CORINFO_HELP_BOX_NULLABLE)
-                                {
-                                    // For nullable we're going to fold it to "ldfld hasValue + brtrue/brfalse" or
-                                    // "ldc.i4.0 + brtrue/brfalse" in case if the underlying type is not castable to
-                                    // the target type.
-                                    CORINFO_RESOLVED_TOKEN isInstResolvedToken;
-                                    impResolveToken(codeAddr + 1, &isInstResolvedToken, CORINFO_TOKENKIND_Casting);
+                            }
+                            else if ((foldAsHelper == CORINFO_HELP_BOX_NULLABLE) &&
+                                     ((impStackTop().val->gtFlags & GTF_SIDE_EFFECT) == 0))
```

Comment:
Didn't do it for nullable due to empty diffs for it (would have to add tests as the logic is a bit different).